packages/contracts-sdk/src/implementations/farcaster/FarcasterKeyGatewayContract.ts

```diff
@@ -0,0 +1,248 @@
+  EVMAccountAddress,
+  EVMContractAddress,
+  BlockchainCommonErrors,
+  UnixTimestamp,
+  FarcasterKeyGatewayContractError,
+  FarcasterId,
+  SignedKeyRequestSignature,
+  EncodedSignedKeyRequestMetadata,
+  SignedAddSignature,
+} from "@snickerdoodlelabs/objects";
+import { ParamType, TypedDataField, ethers } from "ethers";
+import { injectable } from "inversify";
+import { ResultAsync, okAsync } from "neverthrow";
+
+import { BaseContract } from "@contracts-sdk/implementations/BaseContract.js";
+import { IEthersContractError } from "@contracts-sdk/implementations/BlockchainErrorMapper.js";
+import {
+  ContractOverrides,
+  IFarcasterKeyGatewayContract,
+  WrappedTransactionResponse,
+} from "@contracts-sdk/interfaces/index.js";
+import {
+  ContractsAbis,
+  SignedKeyRequest,
+} from "@contracts-sdk/interfaces/objects/index.js";
+
+@injectable()
+export class FarcasterKeyGatewayContract
+  extends BaseContract<FarcasterKeyGatewayContractError>
+  implements IFarcasterKeyGatewayContract
+{
+  constructor(
+    protected providerOrSigner: ethers.Provider | ethers.Signer,
+    protected contractAddress: EVMContractAddress,
+  ) {
+    super(
+      providerOrSigner,
+      contractAddress,
+      ContractsAbis.FarcasterKeyGatewayAbi.abi,
+    );
+  }
+
+  public nonces(
+    address: EVMAccountAddress,
+  ): ResultAsync<
+    bigint,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    return ResultAsync.fromPromise(
+      this.contract.nonces(address) as Promise<bigint>,
+      (e) => {
+        return this.generateError(e, "Unable to call nonces()");
+      },
+    );
+  }
+
+  public add(
+    key: string,
+    encodedMetadata: string, // obtained from getEncodedSignedKeyRequestMetadata()
+    overrides?: ContractOverrides,
+  ): ResultAsync<
+    WrappedTransactionResponse,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    // Based on the docs, only keyType and metadataType of 1 is supported at the moment
+    // https://docs.farcaster.xyz/reference/contracts/reference/key-gateway#add
+    return this.writeToContract("add", [1, key, 1, encodedMetadata], overrides);
+  }
+
+  // Called by third party on behalf of an fid
+  // Called by the owner's EW to sign a 'Register' signature
+  // The signature and encodedMetadata for the 'Add' function must first be obtained using the wallet owning the fid via getSignedKeyRequestSignatureAndEncodedMetadata()
+  public addFor(
+    fidOwnerAddress: EVMAccountAddress,
+    keyToAdd: string,
+    encodedMetadata: string,
+    deadline: UnixTimestamp,
+    signature: SignedAddSignature,
+    overrides?: ContractOverrides,
+  ): ResultAsync<
+    WrappedTransactionResponse,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    // Based on the docs, only keyType and metadataType of 1 are supported at the moment
+    // https://docs.farcaster.xyz/reference/contracts/reference/key-gateway#addfor
+    const keyGatewayAddForParams = {
```

Comment:
Why create this object here? It's not needed, since you just pass the individual params from this object below.